### PR TITLE
feat: Karatsuba algorithm

### PR DESCRIPTION
**Pull request by Swastyy**
#### Description of Change Karatsuba algorithm for fast multiplication
Fixes issue #1507 

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1550"></a>

### CODE REVIEW

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/pull/1550/checks?check_run_id=3201622943) warnings.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/Swastyy/C-Plus-Plus/actions

File: divide_and_conquer/karatsuba_algorithm_for_fast_multiplication.cpp [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+ * Simple approach is to multiply bits one by one which will give the time complexity of 
+ * around O(n^2). To make it more efficient we will be using Karatsuba' algorithm to find the product which will solve the problem O(nlogn) of time.
+ * @author [Swastika Gupta](https://github.com/Swastyy)
+ */
+
+#include <cassert>   /// for assert
+#include <iostream>  /// for IO operations
+#include <vector>    /// for std::vector
+#include <string>    /// for string
+#include <cstring>   /// for string
```
Comment (Panquesito7) [inline divide_and_conquer/karatsuba_algorithm_for_fast_multiplication.cpp:15]: What's the difference between these two? Why not only use one?

File: divide_and_conquer/karatsuba_algorithm_for_fast_multiplication.cpp [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        temp = std::to_string(sum);
+        temp += result;
+        result = temp;
+
+        carry = (firstBit & secondBit) | (secondBit & carry) |
+                (firstBit & carry);  // sum of 3 bits
+    }
+
+    if (carry)
+        result = '1' + result;  // adding 1 incase of overflow
```
Comment (Panquesito7) [inline divide_and_conquer/karatsuba_algorithm_for_fast_multiplication.cpp:73]:
```suggestion
    if (carry) {
        result = '1' + result;  // adding 1 incase of overflow
    }
```

File: divide_and_conquer/karatsuba_algorithm_for_fast_multiplication.cpp [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    // 1st test
+    std::string s11 = "1";
+    std::string s12 = "1010";
+    std::cout << "Test 1... ";
```
Comment (Panquesito7) [inline divide_and_conquer/karatsuba_algorithm_for_fast_multiplication.cpp:135]:
```suggestion
    std::cout << "1st test... ";
```

File: divide_and_conquer/karatsuba_algorithm_for_fast_multiplication.cpp [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    assert(divide_and_conquer::karatsuba_algorithm::karatsuba_algorithm(
+               s11, s12) == 10);  // here the multiplication is 10
+    std::cout << "passed" << std::endl;
+
+    // 2nd test
+    std::string s21 = "11";
+    std::string s22 = "1010";
+    std::cout << "Test 2... ";
```
Comment (Panquesito7) [inline divide_and_conquer/karatsuba_algorithm_for_fast_multiplication.cpp:143]:
```suggestion
    std::cout << "2nd test... ";
```

File: divide_and_conquer/karatsuba_algorithm_for_fast_multiplication.cpp [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    assert(divide_and_conquer::karatsuba_algorithm::karatsuba_algorithm(
+               s21, s22) == 30);  // here the multiplication is 30
+    std::cout << "passed" << std::endl;
+
+    // 3rd test
+    std::string s31 = "110";
+    std::string s32 = "1010";
+    std::cout << "Test 3... ";
```
Comment (Panquesito7) [inline divide_and_conquer/karatsuba_algorithm_for_fast_multiplication.cpp:151]:
```suggestion
    std::cout << "3rd test... ";
```

**Review by Panquesito7 — APPROVED** [review]
LGTM. Thank you for your contribution! 😄👍🎉

**Review by ayaankhan98 — APPROVED** [review]

**Review by mishraabhinn — APPROVED** [review]

**Pull request merged**